### PR TITLE
Propose vote to unblock `--env-file` PRs

### PR DESCRIPTION
**Pull request by aduh95**
Refs: https://github.com/nodejs/TSC/issues/1575#issuecomment-2166976221

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Review by marco-ippolito — APPROVED** [review]

**Review by gireeshpunathil — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 1]
```diff
@@ -24,8 +24,12 @@ headerInstructions: |
...
 #    there are.
 candidates:
-  - TODO
-  - TODO
+  - Throw when passed to `--env-file` flag, provide no other flag (status quo).
+  - Warn when passed to `--env-file` flag, provide no other flag.
+  - Throw when passed to `--env-file` flag, warn when passed to a new `--env-file-if-exists`.
+  - Throw when passed to `--env-file` flag, do nothing when passed to a new `--env-file-if-exists`.
+  - Throw when passed to `--env-file` flag, unless new `--env-file-when-missing=warn` is also passed.
+  - Warn when passed to `--env-file` flag, unless new `--env-file-when-missing=throw` is also passed.
```
Comment (GeoffreyBooth) [inline votes/initiateNewVote/_EDIT_ME.yml:32]: I don't think anyone has been advocating for these. These options would mean you can't have a mix of required and optional files to load.

Comment (aduh95) [reply votes/initiateNewVote/_EDIT_ME.yml:32]: > I don't think anyone has been advocating for these.

I have


> These options would mean you can't have a mix of required and optional files to load.

That's correct

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 # 1. Select a subject for the vote. This can be a question addressed to the TSC
 #    voting members.
-subject: REPLACEME
+subject: What UX do we want `node` to handle non-existent `.env` files passed to its CLI?
```
Comment (tniessen) [inline votes/initiateNewVote/_EDIT_ME.yml:6]: I don't have a suggestion but something in this question seems off, linguistically.

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 3]
```diff
@@ -35,11 +39,15 @@ canShuffleCandidates: true
 # 5. Insert here a short description of the vote objectives and link to the
 #    issue it was discussed on to give the full context.
 footerInstructions: |
-  TBD
+  Based on the result of this vote, we can unblock either https://github.com/nodejs/node/pull/53177,
+  https://github.com/nodejs/node/pull/53060, or neither if the status quo prevails. The name of the
+  proposed flags are non-binding, and are provided only as examples. If a proposal that includes a
+  new flag, the name of that flag can be bike-shedded per the PR review process.
```
Comment (tniessen) [inline votes/initiateNewVote/_EDIT_ME.yml:45]:
```suggestion
  proposed flags are non-binding, and are provided only as examples. If a proposal that includes a
  new flag is accepted, the name of that flag can be bike-shed per the PR review process.
```

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 4]
```diff
@@ -35,11 +39,15 @@ canShuffleCandidates: true
...
 
 # 6. Optionally, insert a brief introduction for the vote PR, in the markdown format.
 prBody: |
+  The vote should stay open for 7 days, or until all TSC voting members have case a ballot.
```
Comment (tniessen) [inline votes/initiateNewVote/_EDIT_ME.yml:49]:
```suggestion
  The vote should stay open for 7 days, or until all TSC voting members have cast a ballot.
```

**Review by tniessen — APPROVED** [review]

**Review by MoLow — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 5]
```diff
@@ -24,8 +24,12 @@ headerInstructions: |
 #    voters express their preference for each candidates, no matter how many
 #    there are.
 candidates:
-  - TODO
-  - TODO
+  - Throw when passed to `--env-file` flag, provide no other flag (status quo).
```
Comment (benjamingr) [inline votes/initiateNewVote/_EDIT_ME.yml:27]: Can we add another option

```suggestion
  - Do nothing (this is the same behavior as `dotenv` and every other library and programming language supporting this).
```

(Assuming there isn't another library or language with different behavior I'm just not aware of)

Comment (benjamingr) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: (I'm also happy with different phrasing)

Comment (anonrig) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: I don't think this is correct. `dotenv` does not throw error if file is not found, but we throw an error right now.

Comment (anonrig) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: Can you rephrase "do nothing" to "don't warn and don't throw an error"?

Comment (aduh95) [reply votes/initiateNewVote/_EDIT_ME.yml:27]:
```suggestion
  - Throw when passed to `--env-file` flag, provide no other flag (status quo).
  - Ignore when passed to `--env-file` flag, provide no other flag.
```

**Comment by anonrig** [comment]
> I really dislike the project landing a feature it's adopting because it reached consensus in userland and then changing its behavior by committee even though all userland libraries (at least popular ones I found) have the same behavior (which is different from what we're proposing).

Note: I agree with @benjamingr 100%.

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 2 above)
Comment (BoscoDomingo) [reply votes/initiateNewVote/_EDIT_ME.yml:6]: Yes, I agree. To me it should be "How should `node` handle non-existent `.env` files passed via CLI"

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 5 above)
Comment (anonrig) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: I think we should make it clear that: `(this is the same behavior as `dotenv` and other libraries and platforms supporting .env)`

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 2 above)
Comment (aduh95) [reply votes/initiateNewVote/_EDIT_ME.yml:6]:
```suggestion
subject: How should `node` handle non-existent `.env` files passed via CLI flags?
```

**Review by ruyadorno — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 5 above)
Comment (benjamingr) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: I am fine with any phrasing that makes it explicit that this option is the one that isn't "we picked an API that reached maturity through userland and we're sticking with the design our community adopted rather than deciding to change its design by committee".

I honestly don't have a strong opinion about whether we should throw or warn or ignore by default but I trust our community and ecosystem to adopt the option with better developer experience. The fact the ~10 implementations I checked* reached the same conclusion and it's different from all our options is troubling to me.


* (and I may be wrong and there are others who reached a different conclusion)

**Comment by BoscoDomingo** [comment]
> > I really dislike the project landing a feature it's adopting because it reached consensus in userland and then changing its behavior by committee even though all userland libraries (at least popular ones I found) have the same behavior (which is different from what we're proposing).
> 
> Note: I agree with @benjamingr 100%.

I agree as well, we just each approached the problem from a different angle. In the end we both want missing .env files to not throw, I just made it a decision to be taken by the user, whereas you prefer sticking to other tools' existing behaviour, and that's 100% fine too 👍🏼 

Also, the proposal LGTM, thanks @aduh95

**Review by BoscoDomingo — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 5 above)
Comment (benjamingr) [reply votes/initiateNewVote/_EDIT_ME.yml:27]: This is and has been the behavior for a while though you can add a config option (verbose:true or debug:true) in order to change it to warn. At least that is my understanding

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 6]
```diff
@@ -24,8 +24,13 @@ headerInstructions: |
...
+  - Ignore when passed to `--env-file` flag, provide no other flag.
+  - Warn when passed to `--env-file` flag, provide no other flag.
+  - Throw when passed to `--env-file` flag, warn when passed to a new `--env-file-if-exists`.
+  - Throw when passed to `--env-file` flag, do nothing when passed to a new `--env-file-if-exists`.
```
Comment (aduh95) [inline votes/initiateNewVote/_EDIT_ME.yml:31]: For the sake of completeness 
```suggestion
  - Throw when passed to `--env-file` flag, ignore when passed to a new `--env-file-if-exists`.
  - Ignore when passed to `--env-file` flag, throw when passed to a new `--env-file-must-exist`.
```

**Comment by benjamingr** [comment]
Basically my stance isn’t warn, ignore or error - it’s “do whatever userland does”.

File: votes/initiateNewVote/_EDIT_ME.yml [hunk 7]
```diff
@@ -35,11 +40,15 @@ canShuffleCandidates: true
 # 5. Insert here a short description of the vote objectives and link to the
 #    issue it was discussed on to give the full context.
 footerInstructions: |
-  TBD
+  Based on the result of this vote, we can unblock either https://github.com/nodejs/node/pull/53177,
+  https://github.com/nodejs/node/pull/53060, or neither if the status quo prevails. The name of the
+  proposed flags are non-binding, and are provided only as examples. If a proposal that includes a
+  new flag is accepted, the name of that flag can be bike-shed per the PR review process.
```
Comment (aduh95) [inline votes/initiateNewVote/_EDIT_ME.yml:46]:
```suggestion
  new flag is accepted, the name of that flag can be bike-shed per the PR review process.
  "Ignore" means "do not throw and do not warn" in this context.
```

**Review by mhdawson — APPROVED** [review]
LGTM

**Review by jasnell — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml (hunk 1 above)
Comment (GeoffreyBooth) [reply votes/initiateNewVote/_EDIT_ME.yml:34]: Sorry I missed that. I feel like a bunch of people would be upset if this is what we land, because the use case they want to support is “I have some env files that every environment needs, plus another env file that only localhost needs”. That was what prompted `--env-file-optional` if I remember correctly.

Comment (aduh95) [reply votes/initiateNewVote/_EDIT_ME.yml:34]: > I feel like a bunch of people would be upset if this is what we land

Yes probably, and that applies to all the candidates BTW – that's why we're having a vote, because we couldn't find a solution that would please everybody.

**Comment by GeoffreyBooth** [comment]
> Basically my stance isn’t warn, ignore or error - it’s “do whatever userland does”.

According to what you’ve written, though, `dotenv` has a default behavior that can be overridden by its config file. We don’t have config file support, though, so at the moment we can’t really mimic `dotenv`‘s behavior, right?

I guess that could be another option, though: add [support for a general Node config file](https://github.com/nodejs/node/pull/53177#issuecomment-2146198245), and then `--env-file` has the warn behavior by default and it can be set to either throw or be silent via the config file. That would exactly match `dotenv`, I think?

**Comment by benjamingr** [comment]
@GeoffreyBooth the equivalent would be "do not warn on missing .env file, add an option to process.loadEnvFile to warn/throw on a missing env file

**Comment by GeoffreyBooth** [comment]
> the equivalent would be “do not warn on missing .env file, add an option to process.loadEnvFile to warn/throw on a missing env file

That’s not equivalent to `--env-file-optional` or `--env-file-required` because by the time you can run `process.loadEnvFile` or `util.parseEnv`, Node is already running and therefore any `NODE_OPTIONS` within the `.env` file has no effect. Part of the existing use case for `--env-file` is as a way to define Node configuration via `NODE_OPTIONS`.

**Comment by benjamingr** [comment]
> That’s not equivalent to --env-file-optional or --env-file-required because by the time you can run process.loadEnvFile or util.parseEnv, Node is already running and therefore any NODE_OPTIONS within the .env file has no effect.

Correct, I was only explaining what userland does and was suggesting we keep the same existing behavior that reached consensus in our ecosystem over many years rather than change it based on how a technical committee believes things should work.
